File: debug/src/main/java/com/sofar/debug/performance/monitor/ThreadMonitor.java
package com.sofar.debug.performance.monitor;

import java.util.Set;

import android.util.Log;

import com.sofar.debug.performance.PerformanceMonitorManager;

public class ThreadMonitor extends AbsMonitor {

  private static final int DUMP_INTERVAL = 10 * 1000;
  private static final String TAG = "ThreadMonitor";
  private static final int WARNING_COUNT = 400;

  private int mLastThreadCount;

  private Runnable runnable = new Runnable() {
    @Override
    public void run() {
      dumpThread();
      record();
      handler.postDelayed(this, DUMP_INTERVAL);
    }
  };

  private void dumpThread() {
    Set<Thread> threadSet = Thread.getAllStackTraces().keySet();
    mLastThreadCount = threadSet.size();
    if (threadSet.size() > WARNING_COUNT) {
      Log.d(TAG, "！警告，当前线程数过多 " + mLastThreadCount);

      StringBuilder sb = new StringBuilder();
      for (Thread t : threadSet) {
        sb.append(t.getName())
          .append("\t");
      }
      Log.d(TAG, "doDump: current threads : \r\n" + sb.toString());
    }
  }


  @Override
  protected void onStart() {
    handler.postDelayed(runnable, DUMP_INTERVAL);
  }

  @Override
  protected void onStop() {
    handler.removeCallbacks(runnable);
  }

  private void record() {
    Log.d(TAG, "当前线程数=" + mLastThreadCount);
    PerformanceMonitorManager.get().threadCallback(mLastThreadCount);
  }

}

File: tool/profiler/src/main/java/com/sofar/profiler/memory/ThreadMonitor.kt
package com.sofar.profiler.memory

import android.util.Log
import com.sofar.profiler.AbsMonitor
import com.sofar.profiler.MonitorManager
import com.sofar.profiler.MonitorType
import java.io.File
import java.lang.Exception

class ThreadMonitor : AbsMonitor() {

  private var tag = "ThreadMonitor"
  private var warningCount = 200

  private var count = 0

  private val runnable: Runnable = object : Runnable {
    override fun run() {
      record()
      handler.postDelayed(this, pollInterval())
    }
  }

  override fun onStart() {
    handler.removeCallbacks(runnable)
    handler.postDelayed(runnable, pollInterval())
  }

  override fun onStop() {
    handler.removeCallbacks(runnable)
  }

  override fun type(): MonitorType {
    return MonitorType.THREAD
  }

  private fun record() {
    try {
      dumpThread()
      Log.d(tag, "thread count=$count")
      MonitorManager.threadCallback(count)
    } catch (e: Exception) {
      Log.d(tag, e.toString())
    }
  }

  fun getCount(): Int {
    return count
  }

  private fun dumpThread() {
    var taskDir = File("/proc/self/task")
    count = taskDir.listFiles().size
    if (count > warningCount) {
      Log.d(tag, getInfo())
    }
  }


  fun getInfo(): String {
    val taskDir = File("/proc/self/task")
    val sb = StringBuilder()
    if (taskDir.exists() && taskDir.isDirectory()) {
      taskDir.listFiles()?.forEach { threadDir ->
        val tid = threadDir.name.toIntOrNull()
        if (tid != null) {
          // 获取线程名称
          val commFile = File(threadDir, "comm")
          if (commFile.exists()) {
            val name = commFile.readText().trim()
            sb.append(name).append("\n")
          }
        }
      }
    }
    return sb.toString()
  }

  fun getJavaThreadInfo(): String {
    val threadSet: Set<Thread> = Thread.getAllStackTraces().keys
    val sb = StringBuilder()
    for (t in threadSet) {
      sb.append(t.name).append("\n")
    }
    Log.d(tag, "thread count={${threadSet.size}}")
    return sb.toString()
  }

}